pkg/genericapiserver/api/handlers/rest.go

```diff
@@ -1079,22 +1113,95 @@ func setListSelfLink(obj runtime.Object, req *restful.Request, namer ScopeNamer)
 	return count, err
 }
 
-func getPatchedJS(patchType types.PatchType, originalJS, patchJS []byte, obj runtime.Object) ([]byte, error) {
+func patchObjectJSON(
+	patchType types.PatchType,
+	codec runtime.Codec,
+	originalObject runtime.Object,
+	patchJS []byte,
+	objToUpdate runtime.Object,
+	versionedObj runtime.Object,
+) (originalObjJS []byte, patchedObjJS []byte, retErr error) {
+	if js, err := runtime.Encode(codec, originalObject); err != nil {
+		return nil, nil, err
+	} else {
+		originalObjJS = js
+	}
 	switch patchType {
 	case types.JSONPatchType:
 		patchObj, err := jsonpatch.DecodePatch(patchJS)
 		if err != nil {
-			return nil, err
+			return nil, nil, err
+		}
+		if patchedObjJS, err = patchObj.Apply(originalObjJS); err != nil {
+			return nil, nil, err
 		}
-		return patchObj.Apply(originalJS)
 	case types.MergePatchType:
-		return jsonpatch.MergePatch(originalJS, patchJS)
+		var err error
+		if patchedObjJS, err = jsonpatch.MergePatch(originalObjJS, patchJS); err != nil {
+			return nil, nil, err
+		}
 	case types.StrategicMergePatchType:
-		return strategicpatch.StrategicMergePatchData(originalJS, patchJS, obj)
+		var err error
+		if patchedObjJS, err = strategicpatch.StrategicMergePatch(originalObjJS, patchJS, versionedObj); err != nil {
+			return nil, nil, err
+		}
 	default:
 		// only here as a safety net - go-restful filters content-type
-		return nil, fmt.Errorf("unknown Content-Type header for patch: %v", patchType)
+		return nil, nil, fmt.Errorf("unknown Content-Type header for patch: %v", patchType)
+	}
+	if err := runtime.DecodeInto(codec, patchedObjJS, objToUpdate); err != nil {
+		return nil, nil, err
+	}
+	return
+}
+
+func strategicPatchObject(
+	codec runtime.Codec,
+	originalObject runtime.Object,
+	patchJS []byte,
+	objToUpdate runtime.Object,
+	versionedObj runtime.Object,
+) (originalObjMap map[string]interface{}, patchMap map[string]interface{}, retErr error) {
+	// TODO: This should be one-step conversion that doesn't require
+	// json marshaling and unmarshaling once #39017 is fixed.
+	data, err := runtime.Encode(codec, originalObject)
+	if err != nil {
+		return nil, nil, err
+	}
+	originalObjMap = make(map[string]interface{})
+	if err := json.Unmarshal(data, &originalObjMap); err != nil {
+		return nil, nil, err
+	}
+	patchMap = make(map[string]interface{})
+	if err := json.Unmarshal(patchJS, &patchMap); err != nil {
+		return nil, nil, err
+	}
+
+	if err := strategicPatchMap(codec, originalObjMap, patchMap, objToUpdate, versionedObj); err != nil {
+		return nil, nil, err
+	}
+	return
+}
+
+func strategicPatchMap(
```

Comment:
godoc, also the name is a little terse for how important it is.  `applyPatchToObject` or something similar?